apps/yggdrasil/test/yggdrasil/character_test.exs

```diff
@@ -0,0 +1,156 @@
+  @missing_assoc_msg "does not exist"
+  @unique_msg "has already been taken"
+  @required_msg "can't be blank"
+
+  setup tags do
+    unless tags[:async] do
+      Ecto.Adapters.SQL.restart_test_transaction(Yggdrasil.Repo, [])
+    end
+
+    :ok
+  end
+
+  setup _context do
+    {:ok, user1} = %User{}
+    |> User.create_changeset(@user1)
+    |> Repo.insert
+
+    {:ok, user2} = %User{}
+    |> User.create_changeset(@user2)
+    |> Repo.insert
+
+    {:ok, game1} = %Game{}
+    |> Game.changeset(@game1)
+    |> Repo.insert
+
+    {:ok, game2} = %Game{}
+    |> Game.changeset(@game2)
+    |> Repo.insert
+
+    {:ok, %{user1: user1, user2: user2, game1: game1, game2: game2}}
+  end
+
+  test "valid character produces a valid changeset", ctx do
+    user = ctx.user1
+    game = ctx.game1
+
+    valid_character = %{@character | user_id: user.id, game_id: game.id}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == true
+  end
+
+  test "valid character changset with valid references inserts", ctx do
+    user = ctx.user1
+    game = ctx.game1
+
+    valid_character = %{@character | user_id: user.id, game_id: game.id}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == true
+
+    {:ok, _character} = Repo.insert character
+  end
+
+  test "invalid character name length < #{@min_len}" do
+    character = Character.changeset(%Character{}, @invalid_character)
+    assert character.valid? == false
+  end
+
+  test "character changset requires game_id" do
+    valid_character = %{@character | user_id: 0}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == false
+    assert {:game_id, @required_msg} in character.errors
+  end
+
+  test "character changset requires user_id" do
+    valid_character = %{@character | game_id: 0}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == false 
+    assert {:user_id, @required_msg} in character.errors
+  end
+
+  test "valid character changset assoc_constraint fires on invalid game", ctx do
+    user = ctx.user1
+
+    valid_character = %{@character | user_id: user.id, game_id: 0}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == true
+
+    {:error, changeset} = Repo.insert character
+    assert {:game, @missing_assoc_msg} in changeset.errors
+  end
+
+  test "valid character changset assoc_constraint fires on invalid user", ctx do
+    game = ctx.game1
+
+    valid_character = %{@character | user_id: 0, game_id: game.id}
+
+    character = Character.changeset(%Character{}, valid_character)
+    assert character.valid? == true
+
+    {:error, changeset} = Repo.insert character
+    assert {:user, @missing_assoc_msg} in changeset.errors
+  end
+
+  test "same character name can be used if the game_id is different with the same user_id", ctx do
```

Comment:
As another example of test names: "it requires the character name to be unique for a game" and "it does not require character names to be unique across games".

If the test names were printed to screen it should be pretty readable. As an example for a full BDD spec printed out:

```
    describe Character
        describe #changeset/2
            it requires character name to be at least 3 characters
            it requires a valid user
            it requires a valid game
            it does not require character name to be unique across games
            it requires character name to be unique inside a game
```